go-controller/pkg/ovn/controller/admin_network_policy/metrics.go

```diff
@@ -0,0 +1,130 @@
+	reg.MustRegister(
+		// expose process metrics like CPU, Memory, file descriptor usage etc.
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+		// expose Go runtime metrics like GC stats, memory stats etc.
+		collectors.NewGoCollector(),
+	)
+}
+
+// Describe is implemented with DescribeByCollect. That's possible because the
+// Collect method will always return the same two metrics with the same two
+// descriptors.
+func (cc ANPControllerCollector) Describe(ch chan<- *prometheus.Desc) {
+	prometheus.DescribeByCollect(cc, ch)
+}
+
+func (cc ANPControllerCollector) fetchANPRuleCountMetric(anpCache map[string]*adminNetworkPolicyState) map[string]map[string]int {
+	ruleCount := map[string]map[string]int{
+		string(libovsdbutil.ACLIngress): {
+			string(anpapi.AdminNetworkPolicyRuleActionAllow): 0,
+			string(anpapi.AdminNetworkPolicyRuleActionPass):  0,
+			string(anpapi.AdminNetworkPolicyRuleActionDeny):  0,
+		},
+		string(libovsdbutil.ACLEgress): {
+			string(anpapi.AdminNetworkPolicyRuleActionAllow): 0,
+			string(anpapi.AdminNetworkPolicyRuleActionPass):  0,
+			string(anpapi.AdminNetworkPolicyRuleActionDeny):  0,
+		},
+	}
+	for _, state := range anpCache {
+		allow, pass, deny := cc.countRules(state.ingressRules)
+		ruleCount[string(libovsdbutil.ACLIngress)][string(anpapi.AdminNetworkPolicyRuleActionAllow)] += allow
+		ruleCount[string(libovsdbutil.ACLIngress)][string(anpapi.AdminNetworkPolicyRuleActionPass)] += pass
+		ruleCount[string(libovsdbutil.ACLIngress)][string(anpapi.AdminNetworkPolicyRuleActionDeny)] += deny
+		allow, pass, deny = cc.countRules(state.egressRules)
+		ruleCount[string(libovsdbutil.ACLEgress)][string(anpapi.AdminNetworkPolicyRuleActionAllow)] += allow
+		ruleCount[string(libovsdbutil.ACLEgress)][string(anpapi.AdminNetworkPolicyRuleActionPass)] += pass
+		ruleCount[string(libovsdbutil.ACLEgress)][string(anpapi.AdminNetworkPolicyRuleActionDeny)] += deny
+	}
+	return ruleCount
+}
+
+func (cc ANPControllerCollector) countRules(rules []*gressRule) (int, int, int) {
+	var passCount, allowCount, denyCount int
+	for _, rule := range rules {
+		switch rule.action {
+		case nbdb.ACLActionAllowRelated:
+			allowCount++
+		case nbdb.ACLActionDrop:
+			denyCount++
+		case nbdb.ACLActionPass:
+			passCount++
+		default:
+			panic(fmt.Sprintf("Failed to count rule type: unknown acl action %s", rule.action))
+		}
+	}
+	return allowCount, passCount, denyCount
+}
+
+// Collect first triggers the fetchANPRuleCountMetric. Then it
+// creates constant metrics for each host on the fly based on the returned data.
+//
+// Note that Collect could be called concurrently, so we depend on
+// fetchANPRuleCountMetric to be concurrency-safe.
+func (cc ANPControllerCollector) Collect(ch chan<- prometheus.Metric) {
+	ruleCount := cc.fetchANPRuleCountMetric(cc.ANPController.anpCache)
+	for direction, actions := range ruleCount {
+		for action, count := range actions {
+			ch <- prometheus.MustNewConstMetric(
```

Comment:
this last bit for my own understanding, how frequently are the pushes? as in how frequently does the "collect" get called?

Reply:
collect is called every time something (e.g. Prometheus) hits the /metrics endpoint.

Reply:
ah I see thanks @simonpasquier and in OCP how frequently do we scrape?

Reply:
the scrape interval is defined by the operator deploying OVN and is probably 30s.

Reply:
I see thanks Simon!!